marshmallow_utils/links.py

```diff
@@ -38,75 +69,101 @@ class LinksStore:
-
-            # Why is this a list?
-            for keyed_parameters in keyed_parameters_array:
-                # Always just a list of 1 dict with
-                # keys: each link key
-                # values: dict with 1
-                #     key: "params"
-                #     value: dict of querystring parameters
-                for k, params in keyed_parameters.items():
-                    keyed_parameters[k] = self.to_link(
-                        config[ns][k],
-                        {**context, **params}
-                    )
-
-    def to_link(self, template, values):
-        """Expand the link template with the values.
-
-        NOTE: querystring parameters (``params``) are converted to associative
-              arrays in order to be expanded correctly.
+        # Iterate over all namespaces
+        for namespace, links_objects in self._links.items():
+            if namespace not in config:
+                raise Exception(f'Unknown links namespace: {namespace}')
+
+            # For each namespace, iterate over the list of link dicts
+            for links in links_objects:
+
+                # Resolve all links in the link dict.
+                rmkeys = []
+                for link_key, link_params in links.items():
+                    try:
+                        template = config[namespace][link_key]
+                        links[link_key] = self.expand_link(
+                            template, {**context, **link_params})
+                    except KeyError:
+                        # no template found - ignore or raise
+                        if self._ignore_missing:
+                            rmkeys.append(link_key)
+                        else:
+                            raise
+
+                # Remove keys that wasn't rendered
+                for k in rmkeys:
+                    del links[k]
+
+    def expand_link(self, template, vars):
+        """Expand the link template with the template variables.
+
+        :param template: URITemplate to expand.
+        :param vars: The template variables for the expansion.
         """
-        qs_dict = values.get("params", {})
-        qs_associative_array = []
-        for param, value in sorted(qs_dict.items()):
-            if isinstance(value, list):
-                param_array = [(param, e) for e in value]
-            else:
-                param_array = [(param, value)]
+        return self.base_url(
+            host=self.host,
+            rendered_path=template.expand(**self.preprocess_vars(vars))
+        )
 
-            qs_associative_array.extend(param_array)
-
-        path = template.expand({**values, "params": qs_associative_array})
-
-        return self.base_url(host=self.host, path=path)
+    @staticmethod
+    def preprocess_vars(vars):
+        """Preprocess template variables before expansion."""
+        for k, v in vars.items():
+            if isinstance(v, MultiDict):
```

Comment:
We probably won't be encountering those too much (schemas playing interference and converting them away), but sure.